Translatable XML help screens

djsrv opened the pull request:
This PR reworks help screens so that they can be more easily translated to other languages. Help screens are now written in human-readable (and writable!) XML and are rendered to PNGs by a new "Help translator" library. Documentation on the [translation workflow](https://github.com/djsrv/snap/blob/help-screens/help/TRANSLATING.md) and the [XML format](https://github.com/djsrv/snap/blob/help-screens/help/XML.md) are in the help folder.

I think this is in a pullable state right now, but it's still missing a few help screens that are exceedingly complicated or require features I haven't yet implemented:
- addCustomBlock
- addVariable
- doFor
- doRun
- evaluateCustomBlock
- reportCombine

This falls back to the old help screens when the new ones aren't present, so that shouldn't be a major problem for the time being.

Due to the sheer number of help screens I had to rewrite in XML, there are probably some errors somewhere. So I need to proofread them all again, and they could probably use a few more eyes on them as well. :)

djsrv committed:
Implement basic HelpDialogMorph

djsrv committed:
Add ScrollFrame to HelpDialogMorph

djsrv committed:
Implement basic help screen deserialization

djsrv committed:
Implement paragraph

djsrv committed:
Use BlockEditorMorph fixLayout in HelpDialogMorph

djsrv committed:
Fetch XML from help folder

djsrv committed:
Implement column/row layout

djsrv committed:
Display scripts in help screens

djsrv committed:
Improve help screen element positioning

djsrv committed:
Allow inline blocks in help screen paragraphs

djsrv committed:
Adjust help screen elements to fill width

djsrv committed:
Fix bugs with help screen element widths

djsrv committed:
Implement basic script diagrams

djsrv committed:
Minor syntax tweaks

djsrv committed:
Remove unnecessary console.log

djsrv committed:
Add padding between block and arrow

djsrv committed:
Fix comment widths

djsrv committed:
Implement help screen images

djsrv committed:
Temporary hack to fix help screen scrolling

djsrv committed:
Add block thumbnail in top-left of help screen

djsrv committed:
Call changed() on help screen images after load

djsrv committed:
Change help screen padding

djsrv committed:
Add forward and bounceOffEdge screens

djsrv committed:
Add bubble help screen

djsrv committed:
Add more help screens

djsrv committed:
Allow gaps in annotation positioning

djsrv committed:
Make arrows originate from annotation's first line

djsrv committed:
Add reverse arrows in diagrams

djsrv committed:
Add custom block support to help screens

djsrv committed:
Fix help screen annotation positioning

Base position on center of first line, not center of the entire morph

djsrv committed:
Add help screens

djsrv committed:
Add menu support to help screens

djsrv committed:
Add more help screens

djsrv committed:
Add script diagram highlights and improved arrows

djsrv committed:
Add result bubbles to script diagrams

djsrv committed:
Allow images in result bubbles

djsrv committed:
Allow morph position to be specified manually

djsrv committed:
Fixes to help screen layout

djsrv committed:
Colored text in help screens

djsrv committed:
Allow Booleans in help screen result bubbles

djsrv committed:
Move more help screen code to help.js

djsrv committed:
Move help code to libraries

djsrv committed:
Complete Help translator library

djsrv committed:
Add help screen sources

djsrv committed:
Merge branch 'master' into help-screens

djsrv committed:
Move old help files

djsrv committed:
Update Help translator library

Export screens as [language].zip

djsrv committed:
Add new rendered help screens

djsrv committed:
Avoid caching help screen sources

djsrv committed:
Add guide on translating help screens

djsrv committed:
Fix script diagram fixLayout

djsrv committed:
Allow blocks to be ghosted in help screens

djsrv committed:
Add more help screens

djsrv committed:
Add black box for doWarp help screen

djsrv committed:
Add script scaling to help screens

djsrv committed:
Prevent arrow overlap in diagrams

djsrv committed:
Help screen cleanup

djsrv committed:
Allow list inputs to be annotated in diagrams

djsrv committed:
Add error handling to help screen blocks

djsrv committed:
Add more help screens

djsrv committed:
Merge branch 'master' into help-screens

djsrv committed:
Fix a comment

djsrv committed:
Add shadows to bubbles in diagrams

djsrv committed:
Fix a bug with bubbles in diagrams

Bubbles would wreck the positioning of menus and annotations

djsrv committed:
Update help screens

djsrv committed:
Add documentation of help screen XML

djsrv committed:
Lint and license help.js

djsrv committed:
Fix XML.md

DyslexicAwe commented:
As a non-English user, I hope it will be tagged with the 'blessed by Jens' tag soon.

jmoenig commented:
very nice, thank you, Dylan @djsrv ! It'll take me a while to check this out and review the changes, but I'm planning to include this for the 5.1 release together with other localization improvements, especially @swiperthefox 's work on supporting better text-editing for accented and non-Western characters.

jmoenig commented:
marking this "blessed by jens" so I don't forget about pulling it :-)

brianharvey commented:
I don't have to review it; Dylan has been showing it to me as he was developing it.  We know that it needs its own documentation, and that there are half a dozen help screens that don't fit into his system to be figured out.  But neither of those should stand in the way of pulling the code (and getting our German and Catalan friends to try making translations).

brianharvey commented:
I don't have to review it; Dylan's been showing it to me as he was developing it.  We know it needs its own documentation, and there are half a dozen help screens it can't handle, but neither of those stands in the way of pulling it.  (And then we can ask our German and Catalan friends to try using it!)

brianharvey commented:
... but I can't figure out how to leave a review so it'll stop asking!  :-(

jmoenig commented:
I'm having a hard time reviewing this as I cannot see the JS files he's changed. This PR has so many files that they don't even show up in the table of contents. I don't need to review the PNGs or XMLs, but the JS.

jmoenig commented:
removing "blessed by Jens" because of this.

brianharvey commented:
There are no changed JS files.  There is a library to be used by translators.  You had said we couldn't change Snap! itself!  Oh, except that the path for helpscreens changes from help/foo.png to help/en/foo.png, help/de/foo.png, etc.

brianharvey reviewed (dismissed):
I don't have to review it; Dylan has been showing it to me as he was developing it. We know that it needs its own documentation, and that there are half a dozen help screens that don't fit into his system to be figured out. But neither of those should stand in the way of pulling the code (and getting our German and Catalan friends to try making translations).

brianharvey commented:
The changed js files are help.js (new), blocks.js, gui.js, and widgets.js.

djsrv commented:
> We know that it needs its own documentation

I've already added that in the help folder. :)

> The changed js files are help.js (new), blocks.js, gui.js, and widgets.js.

Note that help.js and JSZip (which is needed to download all the help screens at once) are only loaded when the library is imported. The changes in the other JS files are just some minor tweaks and a fix to the error handling in the async version of IDE_Morph's getURL.

jmoenig commented:
Okay, thanks, I can look at those JS files now.

cycomachead commented:
First off: This is amazing and I think a huge benefit. 🎆  It would be / will be great to extract these into other places, like a compiled PDF or web page. :)

I'll publish my notes in a minute, but the main post references reviewing all the content for each help screen, so that was why I tagged Brian. By and large, there's nothing major, but a few things to address. Awesome work!! 💯 

For the future, I think something like this is probably most easily reviewed in steps. The first can be code with a couple examples, and then the next set of PRs can be each group of help screens. I also tagged Brian because I think it's good to understand the limitations of the programatic approach and wanted to make sure he was good with that -- though I too don't see anything that should stop this.

cycomachead reviewed (changes requested):
Wow! This is amazing. 

Most of my comments are fairly small details. From testing, I can't seem to scroll when I'm inside the help screen. That's the only real issue I've noticed so far. 

There are a couple minor things, I don't think we need to handle, but are putting out here:
* Should we italicize Snap<em>!</em> properly in help screens? Seems like we should at some point. 
* It would be cool if you could keep open multiple help screens.

cycomachead commented:
Also, we should do this later and not hold this up, but I was initially confused by the goal here. 

Why bother exporting the help screens as an image? Why not load the XML and render it inline? It seems to me like the help renderer functions should at some point become a part of Snap<em>!</em>, and that would make the process of translating slightly easier and give us more "native" text in Snap, which will be useful when we need to better handle screen readers.

brianharvey commented:
Wow, thank you for the extremely detailed review!  I'm in awe of your effort to make this perfect.

I had only one disagreement with your suggestions:  Will a translator necessarily know what a "block spec" is in the Snap<i>!</i> technical sense?  In the manual the user-visible thing in the Block Editor's hat block is called the "prototype" of the block.

As to why not generate the help on the fly instead of storing pictures, you know how Jens is about children writing code inside Snap<i>!</i>, so we were trying hard to minimize that.  Only the translator gets a slightly modified version.  But if Jens wants to accept these changes in the regular version everyone uses, JIT picture generation would be an obvious thing to suggest.

brianharvey commented:
P.S.  Sadly, as of this summer Dylan is no longer a child, so maybe that'll make Jens happier.  :-)

cycomachead commented:
You’re right about the word “spec” being new but I think we need an example or explanation there. Only the spec will work in that place and adding synonyms gets confusing. 
Though if there’s some better term - I’m open to that. We use the spec for the full spec with inputs so it’s already not the best name. 

-- 
Michael Ball
From my iPhone
michaelball.co

> On Jul 29, 2019, at 3:35 AM, Brian Harvey <notifications@github.com> wrote:
> 
> P.S. Sadly, as of this summer Dylan is no longer a child, so maybe that'll make Jens happier. :-)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

cycomachead commented:
Other thoughts:

I should add that the way we should probably handle things is to use Reviews rather than labels. Obviously, Jens' review is the one that matters most, but I think we could make good use of GitHub's features, which do help cleanup labels and it's (relatively) clear when things need more work. 

As far as pulling this into Snap!. I'm really quite impressed with the help.js library. There's definitely a few minor things and I haven't run the serializer changes through every condition, but I think we could get it to a point where it's a part of Snap<em>!</em> for not tons of effort. (Though, again, I do think that's a worthy step 2.)

cycomachead reviewed (commented):
More comments and thoughts, just posting for now.

djsrv commented on help/en/HELP:
```diff
@@ -0,0 +1,175 @@
+bounceOffEdge
```
This list is used by the "download all help screens" block in the help translator library.

djsrv commented on help/en/catch.xml:
```diff
@@ -0,0 +1,170 @@
+        <block-definition s="catch %&apos;tag&apos; %&apos;action&apos;" type="command" category="control">
+            <header></header>
+            <code></code>
+            <translations>de:fange _ _&#xD;ca:agafa _ _&#xD;es:atrapar _ _&#xD;fr:attrape _ _&#xD;</translations>
```
I agree that this isn't ideal. I think the best solution would be to add an attribute somewhere in the XML that says the help screen requires a certain library to be loaded. If it hasn't already been imported (e.g. "download all help screens" is run in a fresh project), the library would be auto-imported, and the screen would take the blocks from the project's global blocks.

djsrv replied:
Of course, if the renderer became part of Snap! and the screens were rendered on-the-fly when someone clicked help, the auto-import thing (or the current solution of just putting the block's XML inside the screen's) wouldn't be necessary. The only situation in which the screen could be rendered is if the library was already imported and someone right-clicked the block.

cycomachead replied:
Yeah, makes sense. :)

djsrv replied:
Fixed in 5c2a076

djsrv committed:
Capitalize title of TRANSLATING.md

Co-Authored-By: Michael Ball <cycomachead@gmail.com>

djsrv committed:
Apply suggested improvements to TRANSLATING.md

Co-Authored-By: Michael Ball <cycomachead@gmail.com>

djsrv committed:
Replace whitespace instead of split+join in help.js

Co-Authored-By: Michael Ball <cycomachead@gmail.com>

djsrv committed:
Remove or rename variables named "screen"

As suggested by @cycomachead

djsrv committed:
Rename processText to normalizeWhitespace

As suggested by @cycomachead

djsrv committed:
Use ide.resourceURL to get image resource URL

As suggested by @cycomachead

djsrv committed:
Rename TRANSLATING.md to README.md

As suggested by @cycomachead

cycomachead commented:
While some of these may be out of date, we can pull in french translations from here:
https://github.com/jmoenig/Snap/pull/298/files

cycomachead commented on src/blocks.js:
```diff
@@ -2253,11 +2257,15 @@ SyntaxElementMorph.prototype.showBubble = function (value, exportPic, target) {
-        pos,
-        isClickable
-    );
+    if (!forHelpScreen) {
```
```suggestion
    if (forHelpScreen) { 
        return bubble;
   }
   bubble.popUp(
       wrrld,
       pos,
       isClickable
   );
```

IMO, this is a bit more clear as to when the function should return early.

cycomachead commented on src/blocks.js:
```diff
@@ -3327,17 +3336,50 @@ BlockMorph.prototype.showHelp = function () {
-            help
-        );
-    };
+    if (lang === 'old') {
```
Why do we need a separate code path for the old/not updated images? Can't we just display them in the same scroll frame? It doesn't seem to me like they work any differently, and if we can do that, then we can have just one set of images, and simplify the error handling function.

djsrv replied:
Yeah, they could probably be combined. If I remember correctly, I kept the old popup code path because:
- The new images are double-sized and scaled down so they don't look terrible on retina displays, while the old ones aren't.
- The new images are all 572 px wide, while the old ones are variable width.
- The new images have padding to make room for a scroll bar on the right side, while the old ones don't.

The first two should be easy to solve. For the last, I'd need to remove the padding from every new image and just add it in when it's drawn in this step.

djsrv replied:
Fixed in 283a82c

djsrv committed:
Make early return in showBubble clearer

Co-Authored-By: Michael Ball <cycomachead@gmail.com>

djsrv committed:
Fix to last commit

djsrv committed:
Combine loading of old and new help screens

djsrv committed:
Load blocks from libraries in help screens

djsrv committed:
Fix script-in-script arrows in diagrams

They didn’t display anymore due to a bug in the arrow detour code.

djsrv committed:
Update help screen font with libre alternative

djsrv committed:
Make sure help screens are rendered at double size

djsrv committed:
Fix help screen typos

djsrv commented:
I've switched to Linux and noticed some problems with this PR that occur outside of the Mac ecosystem, so I've pushed a few changes:
- Allow either Baskerville or Libre Baskerville, since Baskerville is only installed by default on Macs.
- Make sure that help screens are rendered at the same size on retina and non-retina screens.

I've also pushed numerous typo fixes, since I didn't notice my spell checker wasn't enabled for XML over the summer. Sorry about that. :/

Since I'm on break now, I also could break this into multiple PRs like @cycomachead suggested, if that would make this easier to pull. :)
> For the future, I think something like this is probably most easily reviewed in steps. The first can be code with a couple examples, and then the next set of PRs can be each group of help screens.

cycomachead commented:
That sounds great Dylan! 👍 from me. 

-- 
Michael Ball
From my iPhone
michaelball.co

> On Dec 21, 2019, at 2:29 PM, Dylan S <notifications@github.com> wrote:
> 
> ﻿
> I've switched to Linux and noticed some problems with this PR that occur outside of the Mac ecosystem, so I've pushed a few changes:
> 
> Allow either Baskerville or Libre Baskerville, since Baskerville is only installed by default on Macs.
> Make sure that help screens are rendered at the same size on retina and non-retina screens.
> I've also pushed numerous typo fixes, since I didn't notice my spell checker wasn't enabled for XML over the summer. Sorry about that. :/
> 
> Since I'm on break now, I also could break this into multiple PRs like @cycomachead suggested, if that would make this easier to pull. :)
> 
> For the future, I think something like this is probably most easily reviewed in steps. The first can be code with a couple examples, and then the next set of PRs can be each group of help screens.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

brianharvey reviewed (commented):
I am reading through the help screens.  My plan is to do all of that, leaving the actual code for Michael to review in parallel.

I've just started, and I have one minor complaint, which if you have time would be a better use of it imho than splitting up the PR.  Look at doGotoObject.png.  The first three options are equally spaced in the menu (of course), but the arrowheads of the annotations aren't.  The second one should point a little higher.  Even better, if it's easy, would be if the entire annotation group were centered, so the middle arrow would be horizontal and the others slightly angled.

I think probably we made a decision somewhere along the line to suppress small changes in angle of the arrows, but now I'm looking at this screen, it looks unprofessional.

Another alternative, to which I'm open, would be to reduce the font size of these annotations (not font size in general) so that all three arrows could be horizontal.  (The arrowheads would also have to be smaller.)

Meanwhile I'll keep reading...

djsrv commented:
Fixed on my local copy (including the centering request). I need to get up early, so I'll clean up the code and push tomorrow.
![doGotoObject-7](https://user-images.githubusercontent.com/16441443/71317133-44487300-244a-11ea-8232-355437e0181a.png)

brianharvey committed:
help screen changes

Most of these reflect recent changes in the actual blocks.  There are three things I'd like Dylan to work on:

changePenHSVA_graph.png:  This picture, used in several help screens, includes annotations as pixels.  Could they be made translatable?

changeScale.png: descender of "g" cut off.  I noticed it on this screen, but it's a general problem.

doSwitchToCostume.png: I added a script, and as a result, the previous row extends past the right edge of the frame.  I thought maybe it was becasuse of retina, but when I disabled retina mode the same thing happened when I compiled the xml.

Thanks!

brianharvey commented:
I just made a PR, but apparently it wasn't treated as a PR to this PR, but as a new one:

https://github.com/jmoenig/Snap/pull/2540

Here's the comment that went with it:

Most of these reflect recent changes in the actual blocks. There are three things I'd like Dylan to work on:

changePenHSVA_graph.png: This picture, used in several help screens, includes annotations as pixels. Could they be made translatable?

changeScale.png: descender of "g" cut off. I noticed it on this screen, but it's a general problem.

doSwitchToCostume.png: I added a script, and as a result, the previous row extends past the right edge of the frame. I thought maybe it was becasuse of retina, but when I disabled retina mode the same thing happened when I compiled the xml.

Thanks!

brianharvey commented:
More help screens edited -- only three palettes to go!

Below are the things I need Dylan to fix.  Several of them are the same problem: scripts end up wider than the window, even if I turn off retina support, and even if I use the scale="0.8" thing.

reportGetSoundAttribute.png: <br/>, more control over arrows
	Also, titles have moved to below the heading block.

reportPenTrailsAsCostume.xml: Script comes out too wide when compiled. 
	Scale="0.6" doesn't affect it.

reportIfElse.png: Script too wide. And no zebra coloring!

fork.png: xml is correct, png is missing a letter in "draw"

evaluate.png:  In the phrase "the input variable function" the word "function" should be in courier font.  (More generally, it'd be nice to be able to put HTML tags for font, color, etc. inside text.)  Oh, and up on top there's supposed to be an arrow between the two blocks.  (There are a few more like this in other blocks.)

doTellTo.xml: The last script had the first (sprite) input to tell empty.  I put in "Sprite(2)" as it should be, but when I compile it, the script is too wide.

reportAskFor.xml: I changed a word in the text, and when I compiled, all the scripts are too wide.

createClone.png:  This time the help is too /tall/, so the bottom edge is clipped.  (I didn't generate this picture; it came this way in the PR.)

djsrv commented:
>changePenHSVA_graph.png: This picture, used in several help screens, includes annotations as pixels. Could they be made translatable?

It would probably be best to turn this into an SVG. Will do.

>changeScale.png: descender of "g" cut off. I noticed it on this screen, but it's a general problem.

This is a problem that Morphic has with Baskerville in general.
<img width="233" alt="Screen Shot 2019-12-22 at 18 44 45" src="https://user-images.githubusercontent.com/16441443/71329137-1f5a0b80-24ef-11ea-896f-e3c7fb03795e.png">
I'll investigate what's causing it, but I'm a bit scared to mess with something as integral as Snap!'s text rendering.

>[Multiple problems with scripts being too large...]

Those blocks are definitely rendering bigger than they are on my computer(s). The obvious problem would be if you have a custom block zoom. Otherwise, I'm not sure...

Also, the `scale` attribute currently only works on scripts within a diagram tag, which is an oversight on my part. I'll fix that.

>reportGetSoundAttribute.png
>, more control over arrows

The changes I'm making for doGotoObject should allow that.

>Also, titles have moved to below the heading block.
>reportIfElse.png: Script too wide. And no zebra coloring!
>fork.png: xml is correct, png is missing a letter in "draw"

These should all be easy fixes.

>evaluate.png: In the phrase "the input variable function" the word "function" should be in courier font. (More generally, it'd be nice to be able to put HTML tags for font, color, etc. inside text.)

This is harder but shouldn't be too bad since I already made it so arbitrary morphs can be inlined.

>Oh, and up on top there's supposed to be an arrow between the two blocks. (There are a few more like this in other blocks.)

This would actually be a huge pain to implement, for rather few screens. I'll try to figure out how to do it elegantly...

>createClone.png: This time the help is too /tall/, so the bottom edge is clipped. (I didn't generate this picture; it came this way in the PR.)

Not seeing this on my end.

djsrv committed:
Add arrow-horizontal attribute for help annotations

djsrv commented:
I've pushed the changes for doGotoObject. You can mark an annotation with `arrow-horizontal="true"`, which forces the annotation to be positioned so that its arrow is horizontal. Any annotations above are placed higher than they would normally be.
```
<annotations>
    <small-p>Move the sprite to...</small-p>
    <small-p>anywhere on the stage</small-p>
    <small-p arrow-horizontal="true">where the mouse is pointing</small-p>
    <small-p>the point (0, 0)</small-p>
    <small-p>another sprite's position</small-p>
</annotations>
```
![doGotoObject](https://user-images.githubusercontent.com/16441443/71330144-22f29000-24f9-11ea-8581-d07d2abd304e.png)

djsrv committed:
Improve behavior of help screen text within rows

djsrv committed:
Don't make help screen text unnecessarily wide

djsrv committed:
Fix zebra coloring in help screen block definitions

towerofnix commented:
This project looks super amazing!! :'o Here are a, um... a little more than a few nitpicks/mistakes/suggestions I had reviewing all of them.

* The help screen for "broadcast and wait" has info on how to use the dropdown, but "broadcast" doesn't also have it. Probably new users will check out the help for "broadcast" first, so should it either be in that screen or in both?

* The continuation screen suggests checking out the Snap! reference manual, but maybe it should have an image showing how to access it from the menubar?

* In the screen for "script variables", the OOP blurb says "new counter" - should there be an embedded image of that block there instead, like the "temp" image in the description above it? ("inherit" references it too.)

* In the screen for "inherit (attribute)", your "after" image has the direction pale, not the x position.

* One could argue that upscaled images, particularly those that are lower resolution screenshots from the days of BYOB, should not be blurred, since they weren't aliased back then and look somewhat badly filtered now. (This is noticeable in screens like "glide", "for each", and "stamp".) But I'm not sure whether that would be good or bad accessibility-wise, which should be the highest priority of course.

* I'm not sure if there's a screen for "say" yet? (No `doSay.png` screen, corresponding to `doThink.png`.) Probably should just be about the same as the "think" screen, and definitely keep the mention of "say ()" to remove the speech bubble.

* The "reset timer" screen uses a "wait (n) secs" block with a predicate inside it instead of "wait until (condition)"! (So does "get timer" - that screen doesn't have the right block in the corner, either.)

* The blocks "get tempo", "set tempo", and "change tempo" all use nearly the same help screen. I haven't looked at much of the technical details of this PR, but it looks like these each use a separate XML file; probably it's much more work than necessary to combine *those*, but in the translation system, is there some kind of guard against accidentally making the screens inconsistent? Maybe a note in the template, `<!-- getTempo, doSetTempo, and doChangeTempo all have the same content -->`? (Same for  "set variable" and "change variable". There are probably other cases of this too, like show/hide variable, and maybe effect blocks, etc — might be more of a hassle to add comments than it's worth, but shrug!)

* The "stop (...)" screen could perhaps use arrows to the dropdown, like in some of the other screens?

* The "switch to costume" screen pointedly has a script description which is very literally the same words as in the script - on the one hand this is a good thing because it's probably an early help screen for readers, but on the other, maybe it isn't *especially* useful information? ("Wait half a second" instead of "Wait 0.5 secs" would be an easy change for instance.) A brief description of what costumes are would probably be appropriate in this screen too.

* The `evaluate_list.png` file is unused. It should probably either be added to the "call" screen or removed.

* The "launch" screen has a typo - "The move-and-dra script" (or the word "draw" is part covered by the image padding).

* IMO the "move 10 steps" screen should mention something about "point in direction" or "turn", because otherwise it's not clear how it's normally used - it may as well be the same as "change x by"!

* The "message" screen references "when I receive any message" - perhaps it should also be vice versa?

* `getPenAttribute.png` seems somewhat incomplete - it's fairly obvious what it does but the ranges diagram could maybe use a summary?

* In the "pen down?" screen, arguably it should be "if *the* sprite's pen is down."

* Nitpick but should there be a line break after "Sprites 2,3,4,5:" in the "go back layers" screen? :P

* The "a new clone of" screen has the exact same content as "create clone of"; while they do basically the same thing, I think it'd be better to highlight the difference with an example of using the reporter form.

* In `receiveInteraction.png`, the "when scripts are stopped" message has an image of a stop sign - but nothing more specific than that. Maybe it should be the stop sign button, and/or a picture of the "stop (all)" block?

* I don't think this needs to be changed; I just think it's funny that, in screens like "ask for", we say "...to fill the empty slot *circled* in green" when the dropdown slot is clearly indicated with a rectangle :)

* "(maybe this one.)" in `reportAspect.png` — 1) the period should be outside the parentheses, and 2) I think this could be a clearer message? Ala "even the sprite running this block".

* In `reportBoolean.png`, for your programming style tip, I believe you meant "report (foo = 0)", not "report (false)"! :P

* In `reportDate.png`, you probably meant "Report second within minute, 0-59" instead of "0-50".

* In `reportGetSoundAttribute.png`, the arrows should probably be pointing to specific options if possible, like in other help screens.

* In the screen for ">", the script doesn't fit the annotation! It should be something like "y position > 150", not "x position > 200".

* Maybe a sentence in the `reportIfElse.png` screen would be fitting, like "This block is useful for making compact block definitions:"? That way the image in the bottom panel isn't on its own.

* Perhaps the help screen for "letter of" should talk about / give examples of the dropdown options?

* Maybe, in the "item of" screen, there should be an "insert (score) at (1) of (top scores)" block, to make the example more complete?

* In `reportMonadic.png`, the script should use "set y to (100 * ...)", not "change y by".

* In `reportMouseY.png`, the block in the corner of the screen is "mouse x", and so is the block in the example script. Those should be changed :P

* In the help screen for "or", you're using the "and" block!

* Does "pen trails" report the trails of *all* sprites together? If so, the message should probably be "...does not include the stage background or *any* sprites themselves".

* In `reportRandom.png`, you probably want "set x to" and "set y to", not "move steps" twice :P

* In `reportRelationTo.png`, the description of the "direction" option is wrong; it reports not how much the sprite must turn, but the direction in which it must face, to be pointing towards the selected sprite.

* In `reportTextSplit.png`, the first example is probably meant to contain "nowhere  man" (two spaces) in the input, not "hello world". Also, the "split by letter" example doesn't have an output list like the others (its image exists but isn't included in the xml/output).

* In `reportTouchingObject.png`, you've used the "not touching" image twice - for both the "not touching" and "touching" labels! :P

* In `reportVideo.png`, you could use an inline image of "point in direction" instead of the all-caps text.

* In many screens, you talk about how some effects, but not all, function for negative numbers too, and generally suggest experimenting. But in the screen for the plain "set effect" block, you don't (just saying 'type -100 to 100' on its own). It's probably worth fitting there!

* In the screen for "show", the block in the corner is "hide".

* In the turn and turn-left screens, you should probably mention "rotate this much *clockwise*" and "anti-clockwise", because young users who aren't familiar with those terms could easily learn them here.

* The screen for "pen up" is based on the one for "pen down", but it leaves the two additional blocks ("pen up" and another "move 10 steps") without descriptions. Those should probably be added ("stop leaving a pen trail", "move again").

* The screen for "y position" has the block "x position" in the corner.

Hope this helps!! :'D

brianharvey commented:
@towerofnix Thanks for the detailed review.  I've been going through them, too, and found some of your list -- but I can't believe I missed the reportBoolean one!  :-/

djsrv commented:
@towerofnix Thanks! It’s hard to notice errors like these after staring at the help screens for so long.


________________________________
From: Brian Harvey <notifications@github.com>
Sent: Friday, December 27, 2019 13:00
To: jmoenig/Snap
Cc: Dylan S; Mention
Subject: Re: [jmoenig/Snap] Translatable XML help screens (#2456)


@towerofnix<https://github.com/towerofnix> Thanks for the detailed review. I've been going through them, too, and found some of your list -- but I can't believe I missed the reportBoolean one! :-/

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/jmoenig/Snap/pull/2456?email_source=notifications&email_token=AD5OAY7VOA22OH54353TCRDQ2ZNFDA5CNFSM4IEFDBRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHXWBPQ#issuecomment-569335998>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AD5OAY3XCKLRQY3E7YNLPITQ2ZNFDANCNFSM4IEFDBRA>.

brianharvey committed:
Complete bh help edits

... but apparently I still have the wrong library so...

brianharvey reviewed (changes requested):
I have just made a PR to this PR, changing 93 files.  With those changes I'm happy, although Dylan should check out Florrie's complaints just in case I missed one.

djsrv committed:
Help screen cleanup

Cleaning up Brian’s changes, fixes to problems caught by Florrie, and
Scratch Cat -> Alonzo replacements

djsrv committed:
Improve help screen element width handling

Scripts can be scaled even when not in diagrams, and diagrams won’t
take up excessive width if there are no annotations.

djsrv committed:
Fix help screen bubble rendering

djsrv committed:
Properly implement nested help boxes and headers

djsrv committed:
Allow custom fonts in help screens

djsrv committed:
Update XML.md

fill-width has been removed.

DyslexicAwe commented:
Below are my changed "Move block help screen" and "Glide block help screen" which I have updated with mentioning their (in-)dependency on direction value. 
![move block help - add the direction block reference](https://user-images.githubusercontent.com/16387241/72536103-e9077600-3879-11ea-8c55-b5d91e4555a9.png)
and
![glide block help - add the direction block reference](https://user-images.githubusercontent.com/16387241/72536131-f45aa180-3879-11ea-839c-cfcfff4451b6.png)
